pennylane/devices/device_api.py

```diff
@@ -323,12 +323,133 @@ def execute_fn(tapes):
             Only then is the classical postprocessing called on the result object.
 
         """
-        if self.supports_derivatives(execution_config) and execution_config.gradient_method in {
-            "best",
-            None,
-        }:
-            return TransformProgram(), replace(execution_config, gradient_method="device")
-        return TransformProgram(), execution_config
+        execution_config = self.setup_execution_config(execution_config)
+        transform_program = self.preprocess_transforms(execution_config)
+        return transform_program, execution_config
+
+    def setup_execution_config(
+        self, config: ExecutionConfig = DefaultExecutionConfig
+    ) -> ExecutionConfig:
```

Comment:
What would you think of having:
```suggestion
    def setup_execution_config(
        self, config: ExecutionConfig = DefaultExecutionConfig, circuit : Optional[QuantumScript] = None
    ) -> ExecutionConfig:
```

Right now we do:
```
        if device.supports_derivatives(config, circuit=tape):
            new_config = device.preprocess(config)[1]
```
Where we only setup the config once we've validated it's supported. This way we could just do:
```
new_config = device.setup_execution_config(config, circuit)
```
And then we could make smart setup choices based on the circuit. What do you think?

Reply:
At a high level, should preprocessing be circuit-agnostic or can it depend on the circuit? If we optionally allow `setup_execution_config` to be called with a circuit, do we do the same for `preprocess_transforms` as well? We could construct different transform programs based on the content of the circuit and whether it has finite shots or not?

Reply:
I'm also wondering why:
```
if self.supports_derivatives(config) and config.gradient_method in ("best", None):
    return replace(config, gradient_method="device")
```
is a part of `preprocess`. Why don't we directly do 
```
if device.supports_derivatives(config, circuit=tape) and condig.gradient_method in ("best", None):
    new_config = replace(config, gradient_method="device")
```
in `_resolve_diff_method`?

Reply:
Because maybe `"best"` for that circuit is `"adjoint"`, `"backprop"`, or `"my_custom_device_derivative"`.

Reply:
I'm mildly conceptually opposed to passing the `circuit` to the execution config function, just because it seems like that's coming from the qnode and not from the device, so in my head its a more expected chain-of-events that the device makes an `ExecutionConfig` and `TransformProgram` that contains all the "demands" of the device, and then then the qnode can update things if it knows something that isn't device-based about what needs to happen.

But also:
> We could construct different transform programs based on the content of the circuit and whether it has shots?

that does really sound like something we might want to do, and maybe, just so that that is an option, we should have both of these functions accept the tape 🤷‍♀️

Reply:
To be honest, in my mind, the preprocessing program (both the preprocessing of the execution config and preprocessing of the circuits) should not depend on the circuit. It should be a static box that comes with the device, that transforms any circuit and execution config into something that the device can run. This to me makes more sense, so I don't really like making them circuit-dependent.

Reply:
My main goal of this would be being able to replace:
```
if device.supports_derivatives(config, circuit=tape):
    new_config = device.setup_execution_config(config)
else:
     new_config = replace(config, use_device_gradient=False)
     new_config = device.setup_execution_config(config)
```
with a single:
```
new_config  = device.setup_execution_config(config, circuit)
```

So we already have an implicit dependence on the tape via validation checks in `supports_derivatives`.

We need a way to be able to say "yes use the device derivatives" when we are only measuring expectation values, but say "no dont use the device derivatives" when measuring probabilities.

Reply:
I guess this would also enable us to handle things like:
```
if config.gradient_method == "backprop" and bool(circuit.shots):
    # raise an error or something
```
to avoid having the `no_sampling` transform in the preprocess program. As well as:
```
if config.mcm_config.mcm_method == "one-shot" and not bool(circuit.shots):
    # raise an error or something
```